phpseclib/Crypt/Base.php

```diff
@@ -1179,16 +1150,113 @@ function disableContinuousBuffer()
 
         $this->continuousBuffer = false;
         $this->changed = true;
+
+        $this->_setEngine();
     }
 
     /**
-     * Encrypts a block
+     * Test for engine validity
+     *
+     * @see Crypt_Base::Crypt_Base()
+     * @param Integer $engine
+     * @access public
+     * @return Boolean
+     */
+    function isValidEngine($engine)
+    {
+        switch ($engine) {
+            case CRYPT_ENGINE_MCRYPT:
+                return $this->cipher_name_mcrypt &&
+                       extension_loaded('mcrypt') &&
+                       in_array($this->cipher_name_mcrypt, mcrypt_list_algorithms());
+            case CRYPT_ENGINE_INTERNAL:
+                return true;
+        }
+    }
+
+    /**
+     * Sets the preferred crypt engine
+     *
+     * Currently, $engine could be:
+     *
+     * - CRYPT_ENGINE_MCRYPT   [fast]
+     *
+     * - CRYPT_ENGINE_INTERNAL [slow]
+     *
+     * If the preferred crypt engine is not available the fastest available one will be used
      *
-     * Note: Must extend by the child Crypt_* class
+     * @see Crypt_Base::Crypt_Base()
+     * @param Integer $engine
+     * @access public
+     */
+    function setPreferredEngine($engine)
+    {
+        switch ($engine) {
+            //case CRYPT_ENGINE_MCRYPT:
+            case CRYPT_ENGINE_INTERNAL:
+                $this->preferredEngine = $engine;
+                break;
+            default:
+                $this->preferredEngine = CRYPT_ENGINE_MCRYPT;
+        }
+
+        $this->_setEngine();
+    }
+
+    /**
+     * Returns the engine currently being utilized
+     *
+     * @see Crypt_Base::_setEngine()
+     * @access public
+     */
+    function getEngine()
+    {
+        return $this->engine;
+    }
+
+    /**
+     * Sets the engine as appropriate
+     *
+     * @see Crypt_Base::Crypt_Base()
+     * @access private
+     */
+    function _setEngine()
+    {
+        switch (true) {
+            case $this->isValidEngine($this->preferredEngine):
+                $this->engine = $this->preferredEngine;
+                break;
+            case $this->isValidEngine(CRYPT_ENGINE_MCRYPT):
```

Comment:
So something like this?:

```php
$candidateEngines = array($this->preferredEngine, CRYPT_ENGINE_MCRYPT, CRYPT_ENGINE_INTERNAL);
foreach ($candidateEngines as $engine) {
    if ($this->isValidEngine($engine)) {
        $this->engine = $this->preferredEngine;
        break;
    }
}
```

It's doing one more isValidEngine() check than is needed, although I suppose there's nothing wrong with that. It's not like it's in a part of the code that needs to be high performance.

Another possibility, that's a little less readable:

```php
$engines = array($this->preferredEngine, CRYPT_ENGINE_MCRYPT);
for ($i = 0; $i < count($engines) && !$this->isValidEngine($engines[$i]); $i++);
$this->engine = $i == count($engines) ? CRYPT_ENGINE_INTERNAL : $engines[$i];
```